File: kernel/relayflowd/src/engine/hn_poller.rs
use anyhow::{Context, Result};
use relayflowd_core::{Clock, Event, RunSpec};
use serde_json::json;

use super::{Engine, EventSubmitOutcome};

const TOP_STORIES_URL: &str = "https://hacker-news.firebaseio.com/v0/topstories.json";
const DEFAULT_STORY_LIMIT: usize = 5;

/// One-shot adapter from Hacker News' top-stories feed to relayflow events.
pub struct HnPoller {
    spec: RunSpec,
    story_limit: usize,
}

impl HnPoller {
    pub fn new(spec: RunSpec) -> Self {
        Self {
            spec,
            story_limit: DEFAULT_STORY_LIMIT,
        }
    }

    pub fn with_story_limit(mut self, story_limit: usize) -> Self {
        self.story_limit = story_limit;
        self
    }

    pub fn poll_once<C: Clock>(
        &self,
        engine: &Engine<C>,
        created_by: &str,
    ) -> Result<Vec<EventSubmitOutcome>> {
        self.poll_once_with(engine, created_by, |url| {
            ureq::get(url)
                .call()
                .context("fetch HN top stories")?
                .into_string()
                .context("read HN top stories response")
        })
    }

    /// Injecting the fetch operation keeps parsing and event submission fully
    /// deterministic in tests without weakening the production fetch path.
    pub fn poll_once_with<C, F>(
        &self,
        engine: &Engine<C>,
        created_by: &str,
        fetch: F,
    ) -> Result<Vec<EventSubmitOutcome>>
    where
        C: Clock,
        F: FnOnce(&str) -> Result<String>,
    {
        let body = fetch(TOP_STORIES_URL)?;
        let story_ids: Vec<u64> =
            serde_json::from_str(&body).context("parse HN top stories response")?;

        story_ids
            .into_iter()
            .take(self.story_limit)
            .map(|id| {
                engine.submit_event(
                    self.spec.clone(),
                    Event {
                        event_type: "hn.story_posted".into(),
                        payload: json!({"id": id, "type": "story"}),
                        key: None,
                    },
                    created_by,
                )
            })
            .collect()
    }
}

File: sdk/src/hn-poller.ts
/**
 * Hacker News -> relayflow events.
 *
 * This lives OUTSIDE `kernel/` deliberately. An earlier version called Hacker
 * News from `kernel/relayflowd` and review rejected it (PR #16, P1): a
 * durable-execution kernel must not own provider-specific product logic or
 * network I/O, or engine availability and dependencies become coupled to an
 * external service. The kernel gained a `ureq` dependency purely to fetch a
 * JSON feed — a clear sign the code was in the wrong place.
 *
 * So the adapter sits on the authoring surface and submits its events through
 * the journal protocol (`event.submit`), which is the same path any other
 * external producer would use. The kernel learns about Hacker News the way it
 * learns about everything else: as an event.
 */

const TOP_STORIES_URL = 'https://hacker-news.firebaseio.com/v0/topstories.json';
const DEFAULT_STORY_LIMIT = 5;

/** Anything that can submit an event through the journal protocol. */
export interface EventSink {
  eventSubmit(spec: unknown, event: { type: string; payload?: unknown; key?: string }): Promise<unknown>;
}

/** Injected so parsing and submission stay deterministic in tests. */
export type Fetcher = (url: string) => Promise<string>;

/**
 * Typed fetch-transport error. Callers (like sdk/src/cli/hn-monitor.ts)
 * `instanceof` this to distinguish transient HN fetch failures from
 * journal failures — avoids message-string matching (fragile cross-module
 * coupling) and covers the shapes fetch() itself throws (network errors,
 * ECONNREFUSED, TypeError('fetch failed')).
 */
export class HnTransientFetchError extends Error {
  constructor(message: string, cause?: unknown) {
    // Use the native ErrorOptions.cause path so stack formatting and
    // downstream inspectors (util.inspect, structured loggers) see it.
    super(message, cause !== undefined ? { cause } : undefined);
    this.name = 'HnTransientFetchError';
  }
}

const defaultFetcher: Fetcher = async (url) => {
  let response: Response;
  try {
    response = await fetch(url);
  } catch (cause) {
    // Network-layer failures from fetch() itself — TypeError('fetch failed'),
    // ECONNREFUSED, DNS lookup failures — are all transient.
    throw new HnTransientFetchError(`HN fetch failed: ${String(cause)}`, cause);
  }
  if (!response.ok) {
    throw new HnTransientFetchError(`HN fetch failed: HTTP ${response.status}`);
  }
  return response.text();
};

export interface PollOptions {
  storyLimit?: number;
  fetcher?: Fetcher;
  createdBy?: string;
}

/**
 * Fetch the top-stories feed once and submit each story as an event.
 *
 * Dedupe is the kernel's job, not ours: the flow's `dedupeKeyTemplate` plus the
 * (flow, subscription, key) claim means submitting the same story twice wakes
 * it once. This function deliberately does not track what it has already seen.
 */
export async function pollHackerNewsOnce(
  spec: unknown,
  sink: EventSink,
  options: PollOptions = {},
): Promise<unknown[]> {
  const storyLimit = options.storyLimit ?? DEFAULT_STORY_LIMIT;
  const fetcher = options.fetcher ?? defaultFetcher;

  const body = await fetcher(TOP_STORIES_URL);

  let storyIds: unknown;
  try {
    storyIds = JSON.parse(body);
  } catch (cause) {
    throw new HnTransientFetchError(`HN top stories response was not JSON: ${String(cause)}`);
  }
  if (!Array.isArray(storyIds)) {
    throw new HnTransientFetchError('HN top stories response was not an array');
  }

  const outcomes: unknown[] = [];
  for (const id of storyIds.slice(0, storyLimit)) {
    outcomes.push(
      await sink.eventSubmit(spec, {
        type: 'hn.story_posted',
        payload: { id, type: 'story' },
      }),
    );
  }
  return outcomes;
}

export const HN_TOP_STORIES_URL = TOP_STORIES_URL;
